electron/Board.js

```diff
@@ -38,42 +31,120 @@ class Board {
 
   sendPacket(id, values) {
     const p = new Packet(id, values);
-    this.port.send(this.address, p.stringify());
+    const buf = p.toBuffer()
+    if(!buf){
+      console.debug(`[id ${id}] Outbound packet's packet data typing definition could not be found. Not sent.`)
+      return false
+    }
+    this.port.send(this.address, buf);
     return true;
   }
 
+  /**
+   * Uses the offset given in the first packet received from the board to calculate subsequent packet arrival times
+   * @param runTime {number} is the received running duration of the board (in ms)
+   * @returns {number} the estimated timestamp at which the packet was sent (in ms)
+   */
+  calculateTimestamp(runTime) {
+    if (this.firstRecvTime < 0) {
+      /* TODO: consider using multiple packet offsets to reduce likelihood of noise causing the first receive time to
+      *   deviate too significantly */
+      this.firstRecvTime = Date.now()
+    }
+    if (this.firstRecvOffset < 0) {
+      this.firstRecvOffset = runTime
+    }
+    return runTime - this.firstRecvOffset + this.firstRecvTime
+  }
```

Comment:
Be careful, because if a board disconnects and reconnects, it might have rebooted in between. Which would mean that the timestamp on the board would be different. So if the board ever disconnects, then on reconnection it should re-sync timestamps.